lease/lease_queue.go

```diff
@@ -53,3 +55,52 @@ func (pq *LeaseQueue) Pop() interface{} {
 	*pq = old[0 : n-1]
 	return item
 }
+
+// LeaseHeap is a lease heap.
+// Only save one item for a lease, `Push` will update the corresponding lease.
+type LeaseHeap struct {
+	m     map[LeaseID]*LeaseWithTime
+	queue LeaseQueue
+}
+
+func newLeaseHeap() *LeaseHeap {
+	return &LeaseHeap{
+		m:     make(map[LeaseID]*LeaseWithTime),
+		queue: make(LeaseQueue, 0),
+	}
+}
+
+func (mq *LeaseHeap) Init() {
+	heap.Init(&mq.queue)
+	mq.m = make(map[LeaseID]*LeaseWithTime)
+	for _, item := range mq.queue {
+		mq.m[item.id] = item
+	}
+}
+
+func (mq *LeaseHeap) Push(item *LeaseWithTime) {
+	if old, ok := mq.m[item.id]; ok {
+		old.time = item.time
+		heap.Fix(&mq.queue, old.index)
```

Comment:
I don't think we need map to keep uniqueness of the lease in the heap.
Lease's revoke is idempotent and repeated lease is very rare.

Reply:
The client have a `KeepAlive` interface, it will `Renew` the lease periodically. If we use `concurrency/session`, the sesssion will constantly generate new item to the heap.